convert-llama-ggml-to-gguf.py

```diff
@@ -109,28 +143,54 @@ def load(self, data, offset):
         # print(n_dims, name_len, dtype, self.dims, self.name, pad)
         return offset - orig_offset
 
-class GGMLV3Model:
+class GGMLModel:
     def __init__(self):
         self.hyperparameters = None
         self.vocab = None
         self.tensor_map = {}
         self.tensors = []
 
     def validate_header(self, data, offset):
-        if bytes(data[offset:offset + 4]) != b'tjgg' or struct.unpack('<I', data[offset + 4:offset + 8])[0] != 3:
-            raise ValueError('Only GGJTv3 supported')
-        return 8
+        magic = bytes(data[offset:offset + 4])
+        if magic == b'GGUF':
+            raise ValueError('File is already in GGUF format.')
+        if magic == b'lmgg':
+            self.file_format = GGMLFormat.GGML
+            self.format_version = 1
+            return 4
+        version = struct.unpack('<I', data[offset + 4:offset + 8])[0]
+        if magic == b'fmgg':
+            if version != 1:
+                raise ValueError(f'Cannot handle unexpected GGMF file version {version}')
+            self.file_format = GGMLFormat.GGMF
+            self.format_version = version
+            return 8
+        if magic == b'tjgg':
+            if version < 1 or version > 3:
+                raise ValueError(f'Cannot handle unexpected GGJT file version {version}')
+            self.file_format = GGMLFormat.GGJT
+            self.format_version = version
+            return 8
+        raise ValueError(f"Unexpected file magic {magic!r}! This doesn't look like a GGML format file.")
+
+    def validate_conversion(self, ftype):
+        if (self.file_format < GGMLFormat.GGJT or self.format_version < 2) and ftype not in (GGMLFType.ALL_F32, GGMLFType.MOSTLY_F16):
+            raise ValueError(f'Quantizations changed in GGJTv2. Can only convert unquantized GGML files older than GGJTv2. Sorry, your {self.file_format.name}v{self.format_version} file of type {ftype.name} is not eligible for conversion.')
+        if (self.file_format == GGMLFormat.GGJT and self.format_version == 2) and ftype in (GGMLFType.MOSTLY_Q4_0, GGMLFType.MOSTLY_Q4_1, GGMLFType.MOSTLY_Q4_1_SOME_F16, GGMLFType.MOSTLY_Q8_0):
+            raise ValueError(f'Q4 and Q8 quantizations changed in GGJTv3. Sorry, your {self.file_format.name}v{self.format_version} file of type {ftype.name} is not eligible for conversion.')
```

Comment:
Lines of code should be at most 120 chars wide, for readability:
```suggestion
        if ($
            (self.file_format < GGMLFormat.GGJT or self.format_version < 2)$
            and ftype not in (GGMLFType.ALL_F32, GGMLFType.MOSTLY_F16)$
        ):$
            raise ValueError($
                'Quantizations changed in GGJTv2. Can only convert unquantized GGML files older than GGJTv2. Sorry, '$
                f'your {self.file_format.name}v{self.format_version} file of type {ftype.name} is not eligible for '$
                'conversion.'$
            )$
        if ($
            self.file_format == GGMLFormat.GGJT and self.format_version == 2$
            and ftype in ($
                GGMLFType.MOSTLY_Q4_0, GGMLFType.MOSTLY_Q4_1, GGMLFType.MOSTLY_Q4_1_SOME_F16, GGMLFType.MOSTLY_Q8_0,$
            )$
        ):$
            raise ValueError($
                f'Q4 and Q8 quantizations changed in GGJTv3. Sorry, your {self.file_format.name}v{self.format_version} '$
                f'file of type {ftype.name} is not eligible for conversion.'$
            )
```

Reply:
Thanks for the feedback.

In this particular case, I'm honestly not sure putting the time into improving the formatting is really worth it. I doubt anyone else is going to be messing with this script, and it's probably not even going to be in the repo in a month or so. But I made some formatting cleanups anyway.